### ISSUE TITLE
SegQueue.len() cannot be called concurrently while changing

### ISSUE THREAD

**Issue by frostyplanet** [crossbeam-rs/crossbeam#523]
I'm using crossbeam-0.7.3,  wrapping crossbeam channel with future, I use two SegQueues to pass waker of the other end. while running with mutliple thread tokio runtime, SegQueue.len() panic while concurrent pop and push happend. It will reproduce every time.

(Have not try crossbeam-queue-0.2.2 because it's not released)

```
mpsc::bounded::tests::test_future_both_8_thread_multi_4tx ... thread 'tokio-runtime-worker' panicked at 'attempt to subtract with overflow', /home/plan/.cargo/registry/src/github.com-9030ffeae4fbd61d/crossbeam-queue-0.2.1/src/seg_queue.rs:423:21
stack backtrace:
   0:     0x556a57d1b1d4 - backtrace::backtrace::libunwind::trace::h90669f559fb267f0
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1:     0x556a57d1b1d4 - backtrace::backtrace::trace_unsynchronized::hffde4e353d8f2f9a
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2:     0x556a57d1b1d4 - std::sys_common::backtrace::_print_fmt::heaf44068b7eaaa6a
                               at src/libstd/sys_common/backtrace.rs:77
   3:     0x556a57d1b1d4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h88671019cf081de2
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x556a57d3c5bc - core::fmt::write::h4e6a29ee6319c9fd
                               at src/libcore/fmt/mod.rs:1052
   5:     0x556a57d17d07 - std::io::Write::write_fmt::hf06b1c86d898d7d6
                               at src/libstd/io/mod.rs:1426
   6:     0x556a57d1d515 - std::sys_common::backtrace::_print::h404ff5f2b50cae09
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x556a57d1d515 - std::sys_common::backtrace::print::hcc4377f1f882322e
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x556a57d1d515 - std::panicking::default_hook::{{closure}}::hc172eff6f35b7f39
                               at src/libstd/panicking.rs:204
   9:     0x556a57d1d201 - std::panicking::default_hook::h7a68887d113f8029
                               at src/libstd/panicking.rs:224
  10:     0x556a57d1db7a - std::panicking::rust_panic_with_hook::hb7ad5693188bdb00
                               at src/libstd/panicking.rs:472
  11:     0x556a57d1d760 - rust_begin_unwind
                               at src/libstd/panicking.rs:380
  12:     0x556a57d3b0d1 - core::panicking::panic_fmt::hb1f3e14b86a3520c
                               at src/libcore/panicking.rs:85
  13:     0x556a57d3b01d - core::panicking::panic::hcdc9f0ba8d71d265
                               at src/libcore/panicking.rs:52
  14:     0x556a57bfc807 - crossbeam_queue::seg_queue::SegQueue<T>::len::hbeb32f87b392151f
                               at /home/plan/.cargo/registry/src/github.com-9030ffeae4fbd61d/crossbeam-queue-0.2.1/src/seg_queue.rs:423
  15:     0x556a57bdc55d - <qs_common_sync::mpsc::bounded::SharedFutureBoth as qs_common_sync::channel::MPSCShared>::clear_send_wakers::hba8f2b7b8fc89b5c
```

**Comment by stjepang** [comment]
I have just published crossbeam-queue v0.2.2. Can you do `cargo update` and try again?

**Comment by frostyplanet** [comment]
I tried and seems ok to me, thanks

**Comment by frostyplanet** [comment]
when will the version of crossbeam-channel be available?

**Comment by taiki-e** [comment]
crossbeam-queue v0.2.2 seems to have been accidentally published based on recent commit. And seems it is the cause of the version error.

I've yanked crossbeam-queue v0.2.2 and published crossbeam-queue v0.2.3 based on correct commit (https://github.com/crossbeam-rs/crossbeam/commit/18363b5b3e063a940db92c54fdb6f4a0538413f4, [tag](https://github.com/crossbeam-rs/crossbeam/releases/tag/crossbeam-queue-0.2.3))

@frostyplanet Could you check if the version error is fixed?

**Comment by frostyplanet** [comment]
sorry I have given up code dependency on currently calling len(), since I was using it as a hint, and no longer believe that's neccessary.
Or you could add some test cases for this
